src/graphnet/models/task/task.py

```diff
@@ -380,66 +386,85 @@ def _forward(self, x: Union[Tensor, Data]) -> Tensor:  # type: ignore
-                         Available through `Model.nb_outputs`
+            flow_layers: A string indicating the flow layer types. See
+             https://thoglu.github.io/jammy_flows/usage/introduction.html
+             for details.
+            target_norm: A normalization constant used to divide the target
+            values. Value is applied to all targets. Defaults to 1000.
+            hidden_size: The number of columns on which the normalizing flow
+            is conditioned on. May be `None`, indicating non-conditional flow.
         """
         # Base class constructor
-        super().__init__(target_labels=target_labels, **task_kwargs)
 
-    def nb_inputs(self) -> int:
-        """Return number of inputs assumed by task."""
-        return len(self._target_labels)
+        # Member variables
+        self._default_prediction_labels = ["nllh"]
+        self._hidden_size = hidden_size
+        super().__init__(**task_kwargs)
+        self._flow = jammy_flows.pdf(
+            f"e{len(self._target_labels)}",
+            flow_layers,
+            conditional_input_dim=hidden_size,
+        )
+        self._initialized = False
+        self._norm = target_norm
 
-    def _forward(self, x: Tensor, jacobian: Tensor) -> Tensor:  # type: ignore
-        # Leave it as is.
-        return x
+    @property
+    def default_prediction_labels(self) -> List[str]:
+        """Return default prediction labels."""
+        return self._default_prediction_labels
+
+    def nb_inputs(self) -> Union[int, None]:  # type: ignore
+        """Return number of conditional inputs assumed by task."""
+        return self._hidden_size
+
+    def _forward(self, x: Optional[Tensor], y: Tensor) -> Tensor:  # type: ignore
+        y = y / self._norm
+        if x is not None:
+            if x.shape[0] != y.shape[0]:
+                raise AssertionError(
+                    f"Targets {self._target_labels} have "
+                    f"{y.shape[0]} rows while conditional "
+                    f"inputs have {x.shape[0]} rows. "
+                    "The number of rows must match."
+                )
+            log_pdf, _, _ = self._flow(y, conditional_input=x)
+        else:
+            log_pdf, _, _ = self._flow(y)
+        return -log_pdf.reshape(-1, 1)
 
     @final
     def forward(
-        self, x: Union[Tensor, Data], jacobian: Optional[Tensor]
+        self, x: Union[Tensor, Data], data: List[Data]
     ) -> Union[Tensor, Data]:
         """Forward pass."""
-        self._regularisation_loss = 0  # Reset
-        x = self._forward(x, jacobian)
+        # Manually cast pdf to correct dtype - is there a better way?
+        self._flow = self._flow.to(self.dtype)
+        # Get target values
+        labels = get_fields(data=data, fields=self._target_labels)
+        labels = labels.to(self.dtype)
+        # Set the initial parameters of flow close to truth
+        # This speeds up training and helps with NaN
+        if self._initialized is False:
```

Comment:
Just a quick note: I noticed that this part could override the learned weights whenever they are loaded from a state dict, which might affect further training or inference. It could be worth adding a global boolean to check if the weights have been loaded.  I believe this could be done[ here](https://github.com/graphnet-team/graphnet/blob/dc7fa4f4515b28b1ea37838402f0c41ee0619ce3/src/graphnet/deployment/deployment_module.py#L65). Otherwise, it might be good at least to pass `and self.training` , which should work when the model is set to evaluation mode for predictions.

Reply:
@jvaracarbonell thanks for pointing this out! I've added the check